ops/main.py

```diff
@@ -42,6 +43,114 @@
 logger = logging.getLogger()
 
 
+@dataclasses.dataclass(frozen=True)
+class _JujuContext:
+    # source: https://juju.is/docs/juju/charm-environment-variables
+    # the HookVars function: https://github.com/juju/juju/blob/3.6/worker/uniter/runner/context/context.go#L1398
+    # only allow accesing these vars because these are the currently used vars in main.py
+
+    # env var JUJU_ACTION_NAME
+    action_name: Optional[str] = None
+
+    # env var JUJU_ACTION_UUID
+    action_uuid: Optional[str] = None
+
+    # env var JUJU_CHARM_DIR
+    charm_dir: Optional[str] = None
+
+    # env var JUJU_DEBUG
+    debug: Optional[str] = None
+
+    # env var JUJU_DEPARTING_UNIT
+    departing_unit: Optional[str] = None
+
+    # env var JUJU_DISPATCH_PATH
+    dispatch_path: Optional[str] = None
+
+    # env var JUJU_MODEL_NAME
+    model_name: Optional[str] = None
+
+    # env var JUJU_MODEL_UUID
+    model_uuid: Optional[str] = None
+
+    # env var JUJU_NOTICE_ID
+    notice_id: Optional[str] = None
+
+    # env var JUJU_NOTICE_KEY
+    notice_key: Optional[str] = None
+
+    # env var JUJU_NOTICE_TYPE
+    notice_type: Optional[str] = None
+
+    # env var JUJU_PEBBLE_CHECK_NAME
+    pebble_check_name: Optional[str] = None
+
+    # env var JUJU_RELATION
+    relation_name: Optional[str] = None
+
+    # env var JUJU_RELATION_ID
+    relation_id: Optional[int] = None
+
+    # env var JUJU_REMOTE_APP
+    remote_app: Optional[str] = None
+
+    # env var JUJU_REMOTE_UNIT
+    remote_unit: Optional[str] = None
+
+    # env var JUJU_SECRET_ID
+    secret_id: Optional[str] = None
+
+    # env var JUJU_SECRET_LABEL
+    secret_label: Optional[str] = None
+
+    # env var JUJU_SECRET_REVISION
+    secret_revision: Optional[int] = None
+
+    # env var JUJU_STORAGE_ID
+    storage_id: Optional[str] = None
+
+    # env var JUJU_UNIT_NAME
+    unit_name: Optional[str] = None
+
+    # env var JUJU_VERSION
+    version: Optional[str] = None
+
+    # env var JUJU_WORKLOAD_NAME
+    workload_name: Optional[str] = None
+
+    @classmethod
+    def from_environ(cls, env: Dict[str, Any]) -> '_JujuContext':
```

Comment:
Yeah, I think you're right this should now be `from_dict`. However, can we make the type a Mapping so that we don't need `dict()` when passing `os.environ`?